staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness_test.go

```diff
@@ -411,32 +386,356 @@ func TestPriorityAndFairnessWithPanicRecoverAndTimeoutFilter(t *testing.T) {
+		const (
+			requestTimeout                                                     = 3 * time.Second
+			userName                                                           = "alice"
+			fsName                                                             = "test-fs"
+			plName                                                             = "test-pl"
+			serverConcurrency, plConcurrencyShares, plConcurrency, queueLength = 1, 1, 1, 1
+		)
+
+		apfConfiguration := newConfiguration(fsName, plName, userName, plConcurrencyShares, queueLength)
+		stopCh := make(chan struct{})
+		controller, controllerCompletedCh := startAPFController(t, stopCh, apfConfiguration, serverConcurrency, requestTimeout/4, plName, plConcurrency)
+
+		firstRequestTimesOutPath := "/request/first/time-out-as-designed"
+		secondRequestEnqueuedPath := "/request/second/enqueued-as-designed"
+		firstReqHandlerCompletedCh, firstReqInProgressCh := make(chan struct{}), make(chan struct{})
+		firstReqRoundTripDoneCh, secondReqRoundTripDoneCh := make(chan struct{}), make(chan struct{})
+		requestHandler := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			expectMatchingAPFHeaders(t, w, fsName, plName)
+
+			if r.URL.Path == firstRequestTimesOutPath {
+				defer close(firstReqHandlerCompletedCh)
+
+				close(firstReqInProgressCh)
+				<-firstReqRoundTripDoneCh
+
+				// make sure we wait until the caller of the second request returns, this is to
+				// ensure that second request never has a chance to be executed (to avoid flakes)
+				<-secondReqRoundTripDoneCh
+
+				// we expect the timeout handler to have timed out this request by now and any attempt
+				// to write to the response should return a http.ErrHandlerTimeout error.
+				if _, err := w.Write([]byte("foo")); err != http.ErrHandlerTimeout {
+					t.Fatalf("Expected error: %#v, but got: %#v", http.ErrHandlerTimeout, err)
+				}
+				return
+			}
+
+			if r.URL.Path == secondRequestEnqueuedPath {
+				// we expect the concurrency to be set to 1 and so this request should never be executed.
+				t.Fatalf("Expected request to be enqueued: %q", secondRequestEnqueuedPath)
+			}
+		})
+		handler := newHandlerChain(t, requestHandler, controller, userName, requestTimeout)
+
+		server, requestGetter := newHTTP2ServerWithClient(handler)
+		defer server.Close()
+
+		wg := sync.WaitGroup{}
+		wg.Add(2)
+		var firstReqErr, secondReqErr error
+		var resp1, resp2 *http.Response
+		go func() {
+			defer wg.Done()
+			defer close(firstReqRoundTripDoneCh)
+			t.Logf("Waiting for the request: %q to time out", firstRequestTimesOutPath)
+			resp1, firstReqErr = requestGetter(firstRequestTimesOutPath)
+		}()
+		func() {
+			defer wg.Done()
+			defer close(secondReqRoundTripDoneCh)
+
+			// we must wait for the "first" request to start executing first
+			<-firstReqInProgressCh
+			resp2, secondReqErr = requestGetter(secondRequestEnqueuedPath)
+		}()
+
+		wg.Wait()
```

Comment:
We do not even need a WaitGroup here, `<- firstReqRoundTripDoneCh` would suffice instead of this `wg.Wait()`, right?

Reply:
> We do not even need a WaitGroup here, <- firstReqRoundTripDoneCh would suffice instead of this wg.Wait(), right?
yes, I removed the WaitGroup.

> In a similar purist vein: we could replace all the channel mucking about with WaitGroup mucking about, right? I do not have a strong preference between them.
Not sure if all the channels can be replaced with a WaitGroup, the test, in different code sites waits for a particular channel. I guess it could be possible if we had a named WaitGroup.

Reply:
A WaitGroup does not have a name per se, but you had one in a variable with a name; different WaitGroups could be held in different variables.